File: mvp/src/main/java/com/shea/mvp/presenter/BasePresenter.java
package com.shea.mvp.presenter;

import android.os.Bundle;

import com.shea.mvp.interactor.BaseInteractor;
import com.shea.mvp.view.BaseView;


public abstract class BasePresenter<I extends BaseInteractor, V extends BaseView> {
    protected I interactor;
    protected V view;

    public BasePresenter(I interactor, V view) {
        this.interactor = interactor;
        this.view = view;
    }

    public final void setupViews(Bundle savedInstanceState) {
        view.setupViews(savedInstanceState);
        onSetupViews(savedInstanceState);
    }

    public final void resume() {
        onResume();
    }

    public final void pause() {
        onPause();
    }

    public final void destroy() {
        view.destroy();
    }

    protected void onResume() { /*Override*/ }
    protected void onPause() { /*Override*/ }
    protected void onSetupViews(Bundle savedInstanceState) { /*Override*/ }
    public void onSaveState(Bundle outState) { /* Override*/ }
}

File: src/main/java/com/shea/mvp/presenter/BasePresenter.kt
package com.shea.mvp.presenter

import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.OnLifecycleEvent
import android.os.Bundle
import com.shea.mvp.repository.BaseRepository


abstract class BasePresenter<I : BaseRepository, V : BaseInterface.BaseViewInterface>(protected var interactor: I, protected var view: V) : BaseInterface.BasePresenterInterface {
    /**
     * This method is called after the View's views are set up, meaning
     * it's safe to reference them.
     */
    open fun onSetupViews(savedInstanceState: Bundle?) { /*Override*/
    }


    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    open fun onResume() {
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    open fun onPause() {
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    open fun onDestroy() {
        view.destroy()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    open fun onCreate() {
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    open fun onStart() {
        view.attach(this)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    open fun onStop() {
        view.detach()
    }

    /**
     * Override this to save instance state
     */
    override fun onSaveState(outState: Bundle) { /* Override*/
    }
}